core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java

```diff
@@ -37,130 +37,147 @@
  */
 abstract class AbstractEventContext implements BaseEventContext {
 
+  private static final int STATE_READY = 0;
+  private static final int STATE_RESPONSE = 1;
+  private static final int STATE_COMPLETE = 2;
+  private static final int STATE_TERMINATED = 3;
+
   private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventContext.class);
-  protected static final FlowExceptionHandler NULL_EXCEPTION_HANDLER = NullExceptionHandler.getInstance();
+  private static final FlowExceptionHandler NULL_EXCEPTION_HANDLER = NullExceptionHandler.getInstance();
+
+  private transient final List<BaseEventContext> childContexts = new ArrayList<>();
+  private transient final FlowExceptionHandler exceptionHandler;
+  private transient final CompletableFuture externalCompletionFuture;
+  private transient final List<BiConsumer<CoreEvent, Throwable>> onResponseConsumerList = new ArrayList<>();
+  private transient final List<BiConsumer<CoreEvent, Throwable>> onCompletionConsumerList = new ArrayList<>();
+  private transient final List<BiConsumer<CoreEvent, Throwable>> onTerminatedConsumerList = new ArrayList<>();
 
-  private transient MonoProcessor<CoreEvent> beforeResponseProcessor;
-  private transient MonoProcessor<CoreEvent> responseProcessor;
-  private transient MonoProcessor<Void> completionProcessor;
-  private transient Disposable completionSubscriberDisposable;
-  private transient final List<BaseEventContext> childContexts = new LinkedList<>();
-  private transient Mono<Void> completionCallback;
-  private transient FlowExceptionHandler exceptionHandler;
+  private volatile int state = STATE_READY;
+  private volatile Either<Throwable, CoreEvent> result;
 
   public AbstractEventContext() {
-    this(NULL_EXCEPTION_HANDLER, empty());
+    this(NULL_EXCEPTION_HANDLER, null);
   }
 
   public AbstractEventContext(FlowExceptionHandler exceptionHandler) {
-    this(exceptionHandler, empty());
+    this(exceptionHandler, null);
   }
 
-  public AbstractEventContext(FlowExceptionHandler exceptionHandler, Publisher<Void> completionCallback) {
-    this.completionCallback = from(completionCallback);
+  public AbstractEventContext(FlowExceptionHandler exceptionHandler, CompletableFuture completableFuture) {
```

Comment:
make the completableFuture optional, or use one that returns a null result instead of nulling the future.
Also, rename the param and add javadoc to have an idea what it is used for